src/include/daos_pipeline.h

```diff
@@ -0,0 +1,256 @@
+ *
+ * \param[in,out]	pipeline	Pipeline object.
+ *
+ * \param[in]		node		Node object to be added to the pipeline.
+*/
+int
+daos_pipeline_push(daos_pipeline_t *pipeline, daos_pipeline_node_t *node);
+
+/**
+ * Adds a new filter object to the pipeline node object 'node'.
+ *
+ * \param[in,out]	node	Pipeline node object.
+ *
+ * \param[in]		filter	Filter object to be added to a pipeline node.
+ */
+int
+daos_pipeline_node_push(daos_node_t *node, daos_pipeline_filter_t *filter);
+
+/**
+ * Checks that a pipeline object is well built. If the pipeline object is well
+ * built, the function will return 0 (no error).
+ *
+ * \param[in]		pipeline	Pipeline object.
+ */
+int
+daos_pipeline_check(daos_pipeline_t *pipeline);
+
+/**
+ * Runs a pipeline on DAOS, returning objects and/or aggregated results.
+ *
+ * \param[in]		pipeline	Pipeline object.
+ *
+ * \param[in]		th		Optional transaction handle. Use
+ *					DAOS_TX_NONE for an independent
+ *					transaction.
+ *
+ * \param[in]		flags		Conditional operations.
+ *
+ * \param[in]		dkey		Optional dkey. When passed, no iteration
+ *					is done and processing isonly performed
+ *					on this specific dkey.
+ *
+ * \param[in]		nr_iods		Number of I/O descriptors in the iods
+ *					table.
+ *
+ * \param[in,out]	iods		[in]: Array of I/O descriptors. Each
+ *					descriptor is associated with a given
+ *					akey and describes the list of
+ *					record extents to fetch from the array.
+ *					[out]: Checksum of each extent is
+ *					returned via
+ *					\a iods[]::iod_csums[]. If the record
+ *					size of an extent is unknown (i.e. set
+ *					to DAOS_REC_ANY as input), then the
+ *					actual record size will be returned in
+ *					\a iods[]::iod_size.
+ *
+ * \param[in,out]	anchor		Hash anchor for the next call, it should
+ *					be set to zeroes for the first call, it
+ *					should not be changed by caller
+ *					between calls.
+ *
+ * \param[in,out]	nr_kds		[in]: Number of key descriptors in
+ *					\a kds.
+ *					[out:] Number of returned key descriptors.
+ *
+ * \param[in,out]	kds		[in]: Optional preallocated array of \nr
+ *					key descriptors.
+ *					[out]: Size of each individual key along
+ *					with checksum type and size stored just
+ *					after the key in \a sgl_keys.
+ *
+ * \param[out]		sgl_keys	Optional sgl storing all dkeys to be
+ *					returned.
+ *
+ * \param[out]		sgl_recx	Optional sgl storing all the records to
+ *					be returned.
+ *
+ * \param[out]		sgl_agg		Optional sgl with the returned value of
+ *					the aggregator(s).
+ *
+ * \param[in]		ev		Completion event. It is optional.
+ *					Function will run in blocking mode if
+ *					\a ev is NULL.
+ */
+int
+daos_pipeline_run(daos_pipeline_t *pipeline, daos_handle_t th, uint64_t flags,
+		  daos_key_t *dkey, uint32_t nr_iods, daos_iod_t *iods,
+		  daos_anchor_t *anchor, uint32_t *nr_kds, daos_key_desc_t *kds,
+		  d_sg_list_t *sgl_keys, d_sg_list_t *sgl_recx,
+		  d_sg_list_t *sgl_agg, daos_event_t *ev);
```

Comment:
I still think oh should be in the api interface to match other object API.

Reply:
Really? So what should be the approach then? I am confused now :-)

Should we include dkey and iods inside the pipeline data structure?  or should we move oh to the API call?

The reason why I moved the object to the pipeline structure is because this is a call over the pipeline, not the object. daos_pipeline_run (instead of daos_obj_pipeline_*)

Reply:
Thanks for your suggestions !

Reply:
I mean this pipeline is just object enumeration + fetch with some extra filter, so the interface should be similar as other object api,  so I assume the interface should be

(obj_oh, th, flags, anchor, iod, sgl, nr,  filter, ....). 

I think it is fine either way whether you choose to put oh into the filter or not. IMHO.

Reply:
**_That being said I do not see how this is different with other object API, maybe I miss sth._**

The only difference will be if in the future we decided that pipelines can work over multiple objects at once (something akin to a join in a DB).

Reply:
@wangdi1 other object API do not do filtering.
and they enumerate and fetch all data to the client side. this is not the purpose here.
the purpose is to ship and apply the filter on the server side before sending the data to the client.

Reply:
I do not have any objections though whether to make  the API similar to object API, or to move those fields to daos_pipeline_t.

Reply:
>The only difference will be if in the future we decide that pipelines can work over multiple objects at once (something akin to a join in a DB).

I am not sure we should combine multiple objects in one API, since you will need dkey and object (layout) to figure out which server to send the pipeline RPC.  i.e. if you do this per object, might  just call this API multiple times?  Though I might miss some use cases here.

>other object API do not do filtering.
and they enumerate and fetch all data to the client side. this is not the purpose here.

Oh, I meant for API interface, after all  it is just special enumeration + fetch for individual object?

Anyway, I do not have strong opinion here.

Reply:
OK. I took the decision of putting oh in the API call and out of the daos_pipeline_t structure.